compiler/next/lib/Queries/Context.cpp

```diff
@@ -0,0 +1,319 @@
+  // but it's technically possible that a platform library
+  // could not do so. So, here we check.
+  // We only need even alignment here.
+  if ((((uintptr_t)buf) & 1) != 0) {
+    free(buf);
+    // try again with an aligned allocation
+    size_t alignment = sizeof(void *);
+    size_t align_up_len = ALIGN_UP(amt, sizeof(void*));
+    buf = (char*) aligned_alloc(alignment, align_up_len);
+  }
+  assert(buf);
+  assert((((uintptr_t)buf) & 1) == 0);
+  return buf;
+}
+
+const char* Context::getOrCreateUniqueString(const char* str) {
+  char gcMark = this->gcCounter & 0xff;
+  auto search = this->uniqueStringsTable.find(str);
+  if (search != this->uniqueStringsTable.end()) {
+    char* buf = search->second;
+    // update the GC mark
+    // Performance: Would it be better to do this store unconditionally?
+    if (this->currentRevisionNumber == this->lastPrepareToGCRevisionNumber) {
+      buf[0] = gcMark;
+    }
+    const char* key = buf+2; // pass the 2 bytes of metadata
+    return key;
+  }
+  size_t strLen = strlen(str);
+  size_t allocLen = strLen+3; // 2 bytes of metadata, str data, '\0'
+  char* buf = allocateEvenAligned(allocLen);
+  // set the GC mark
+  buf[0] = gcMark;
+  // set the unused metadata (need to still have even alignment)
+  buf[1] = 0x02;
+  // copy the string data, including the null terminator
+  memcpy(buf+2, str, strLen+1);
+  const char* key = buf+2; // pass the 2 bytes of metadata
+  // Add it to the table
+  this->uniqueStringsTable.insert(search, {key, buf});
+  return key;
+}
+
+const char* Context::uniqueCString(const char* s) {
+  if (s == nullptr) s = "";
+  return this->getOrCreateUniqueString(s);
+}
+
+UniqueString Context::moduleNameForID(ID id) {
+  // If the symbol path is empty, this ID doesn't have a module.
+  if (id.symbolPath().isEmpty()) {
+    UniqueString empty;
+    return empty;
+  }
+
+  // Otherwise, the module name is everything up to the first '.'
+  size_t len = 0;
+  const char* s = id.symbolPath().c_str();
+  while (true) {
+    if (s[len] == '\0' || s[len] == '.') break;
+    len++;
+  }
+
+  return UniqueString::build(this, s, len);
+}
+
+UniqueString Context::filePathForID(ID id) {
+  UniqueString modName = this->moduleNameForID(id);
+  return this->filePathForModuleName(modName);
+}
+
+UniqueString Context::filePathForModuleName(UniqueString modName) {
+  QUERY_BEGIN_NAMED(this, UniqueString, "filePathForModuleName", modName);
+  if (QUERY_USE_SAVED()) {
+    return QUERY_GET_SAVED();
+  }
+  assert(false && "This query should always use a saved result");
+  auto result = UniqueString::build(this, "<unknown file path>");
+  return QUERY_END(result);
+}
+
+void Context::advanceToNextRevision(bool prepareToGC) {
+  this->currentRevisionNumber++;
+  if (prepareToGC) {
+    this->lastPrepareToGCRevisionNumber = this->currentRevisionNumber;
+    gcCounter++;
+  }
+  printf("CURRENT REVISION NUMBER IS NOW %i\n",
+         (int) this->currentRevisionNumber);
```

Comment:
Should this be in an `if (debug)` of sorts?

Reply:
Oh yeah I'll just remove it for now, thanks.